lib/oneFFS.ml

```diff
@@ -163,3 +163,50 @@ module Make(B : Mirage_block.S) = struct
       Cstruct.blit_from_string Header.empty 0 buf 0 (String.length Header.empty);
       Lwt.return { b; info; f; empty_header = buf; }
 end
+
+module Make_lossy = Make
+
+module Make_lossless(Pclock : Mirage_clock.PCLOCK)(Block : Mirage_block.S) = struct
+  module FS = Filesystem.Make(Pclock)(Block)
+
+  type t = {
+    b : Block.t;
+    mutable superblock : FS.superblock;
+  }
+
+  let connect b =
+    let* info = Block.get_info b in
+    if info.sector_size < FS.superblock_size then
+      failwith "sector size too small for this filesystem";
+    let* r = FS.read_data b in
+    match r with
+    | Error `Bad_checksum ->
+      let* r = FS.init b in
+      (match r with
+       | Ok superblock ->
+         Lwt.return ({ superblock; b }, None)
+       | Error `Msg e ->
+         Printf.ksprintf failwith "error initializing the block device: %s" e)
+    | Error `Msg e ->
+      Printf.ksprintf failwith "error reading block device: %s" e
+    | Error (#FS.decode_err as e) ->
+      Format.kasprintf failwith "error reading block device: %a" FS.pp_decode_err e
+    | Ok (superblock, data) ->
+      let t = { superblock; b } in
+      Lwt.return (t, Some data)
+
+  let write t data =
+    let+ r = FS.write_data t.b t.superblock data in
+    Result.map (fun superblock -> t.superblock <- superblock) r
+
+  let read t =
+    let+ r = FS.read_data t.b in
+    match r with
+    | Ok (superblock, data) ->
+      (* XXX: what to do with the read superblock?! *)
```

Comment:
so in the tlstunnel unikernel, we do:

```OCaml
FS.read_data block >>= function
...
    | Ok (superblock, data) ->
      let config = { superblock ; sni = Domain_name.Host_map.empty } in
...

   let write_configuration block config =
    let open Lwt_result.Infix in
    let data = Configuration.encode_data config.sni in
    FS.write_data block config.superblock data >|= fun superblock ->
    config.superblock <- superblock
```

So, it looks for writing we need the superblock. But we already have that from `connect`... so a read is pretty superfluous?!? i.e. when would you want to read again? certainly you read at startup, and then when you change something you write, but you should never need to read again.

But for the XXX: we can just ignore the superblock :)

Reply:
Yes, exactly! I would only read again if for some reason I decide not to keep the data in memory. But the format doesn't lend well to partial reads, so you wouldn't store more data than you can keep in memory.